File: Questao-4/Questao4.py
def dfs(a, grafo, visitados):
    visitados[a] = 1
    for j in range(N):
        if grafo[a][j]:
            if visitados[j]:
                return True
            if dfs(j, grafo, visitados):
                return True
    visitados[a] = 0
    return False

T = int(input())

while T:
    circulo = 0

    N, M = [int(x) for x in input().split(' ')]
    grafo = [([0]*N).copy() for x in range(N)]
    
    while M:
        A, B = [int(x) for x in input().split(' ')]
        grafo[A-1][B-1] = 1
        M -= 1
    
    visitados = [0] * N
    for i in range(N):
        if dfs(i, grafo, visitados):
            circulo = 1
    print("SIM" if circulo else "NAO")
    T -= 1

File: Questao-4/Questao4.c
#include <stdio.h>
#include <string.h>

int N, circulo;

void dfs(int a, int grafo[N][N], int visitados[N]){
    if (circulo) return;

    visitados[a] = 1;
    for (int j = 0; j < N; j++){
        if (grafo[a][j]){
            if (visitados[j]){
                circulo = 1;
            }
            else{
               dfs(j, grafo, visitados); 
            }
        }
    }
    visitados[a] = 0;
}

int main(){
    int T, M, A, B;

    scanf("%d", &T);
    while(T--){
        circulo = 0;
        scanf("%d %d", &N, &M);

        int grafo[N][N];
        memset(grafo, 0, sizeof(grafo));

        while(M--){
            scanf("%d %d", &A, &B);
            grafo[A-1][B-1] = 1;
        }


        int visitados[N];
        memset(visitados, 0, sizeof(visitados));

        for (int i = 0; i < N; i++){
            dfs(i, grafo, visitados);
            if (circulo) break;
        }
        
        if (circulo) printf("SIM\n");
        else printf("NAO\n");

    }
    return 0;
}